### DISCUSSION TITLE
How to make builds (of forks or older releases) work after a few months?

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#1950] [Q&A] [answered]
We have our own fork with some modifications.

I try to update this to the latest BIRT version once or twice a year or so, but sometimes, the changes in the official BIRT are more than I can test, so I want eg to cherry-pick only a handful of commits.

I know how this can be done with Git and my source code looks the way I want it.

But in many cases, I am not able to build, and I think the reason for this is that the `org.eclipse.birt.target.target` file contains a reference to a repository that is no longer available.

For example, yesterday I performed this work for a branch I last built in February 2024 (and it was up-to-date with master then), and referenced the repository 

      <repository location="https://download.eclipse.org/eclipse/updates/4.31-I-builds"/>

... which no longer exists.

I changed this line to 

      <repository location="https://download.eclipse.org/eclipse/updates/4.31"/>

and had to fix some issues in several MANIFEST.MF files caused by the renaming of `org.apache.commons.codec` to `org.apache.commons.commons-codec` (which seems stupid IMHO, but anyway...) and version conflicts of that package.

This way I was able to run the All-In-One from within Eclipse, but the Maven build still failed (complained about JustJ version):

[ERROR] Cannot resolve target definition:
[ERROR]   Software being installed: org.eclipse.justj.epp.feature.group 21.0.0.v20240807-1553
[ERROR]   Missing requirement: org.eclipse.justj.epp 21.0.0.v20240807-1553 requires 'osgi.bundle; org.eclipse.justj.openjdk.hotspot.jre.full [21.0.4.v20240802-1551,22.0.0)' but it could not be found
[ERROR]   Cannot satisfy dependency: org.eclipse.justj.epp.feature.group 21.0.0.v20240807-1553 depends on: org.eclipse.equinox.p2.iu; org.eclipse.justj.epp [21.0.0.v20240807-1553,21.0.0.v20240807-1553]

Similar problems occur when one tries to build an older BIRT release from source (at least without a Nexus server).

My questions are:

1) Main question: How can I create a build based on a fixed set of dependencies which can still be built several months later as well?

2) Is changing the single line in the `*.target` file like I did the right way?

I guess the suffix `-I-builds` means "intermediate builds on the way to the release version", and without the suffix it is the release version.

3) What can I do about the JustJ problem in the Maven build?

4) Why does Eclipse remove the intermediate repositories at all?

If the file system or repo-software that is used internally deduplicates BLOBs, keeping it (instead of deleting it) should be no problem.

5) How do others tackle this?

6) Would it even be possible to have such reliable builds for BIRT releases as well?

Note: I'm not even talking about _reproducible_ builds, which would mean that the build artefacts are bit-wise identical everytime they are built.

**Comment by merks** [comment, marked as answer]
> We have our own fork with some modifications.
> 

The terms of the EPL require you to make such modifications public.  😨 

> I try to update this to the latest BIRT version once or twice a year or so, but sometimes, the changes in the official BIRT are more than I can test, so I want eg to cherry-pick only a handful of commits.
> 
> I know how this can be done with Git and my source code looks the way I want it.
> 
> But in many cases, I am not able to build, and I think the reason for this is that the `org.eclipse.birt.target.target` file contains a reference to a repository that is no longer available.
> 

Note that for the most recent releases I've been stabilizing the target platform fix repository locations:

https://github.com/eclipse-birt/birt/commit/c41eeb2165cf5f0c5765edc12067d14f33bdfbce

The only repo left with moving content is this one:

https://download.eclipse.org/justj/jres/21/updates/release/latest 

Because I generally expect that no matter when one builds one will want the latest Java release with the latest CVE fixes.

> For example, yesterday I performed this work for a branch I last built in February 2024 (and it was up-to-date with master then), and referenced the repository
> 
In the middle of a release cycle there will be many update sites with content that changes over time.
> ```
>   <repository location="https://download.eclipse.org/eclipse/updates/4.31-I-builds"/>
> ```
> 
> ... which no longer exists.
> 
Yes, such as this one, which is actually deleted after the release.
> I changed this line to
> 
> ```
>   <repository location="https://download.eclipse.org/eclipse/updates/4.31"/>
> ```
> 
> and had to fix some issues in several MANIFEST.MF files caused by the renaming of `org.apache.commons.codec` to `org.apache.commons.commons-codec` (which seems stupid IMHO, but anyway...) and version conflicts of that package.
> 
To ship quality result we should generally try to ship with the latest version of every dependency.  The those dependencies decide to rename their bundles is beyond our control.  Generally using package imports instead of bundle requirements helps avoid that, but the code base how it is.
> This way I was able to run the All-In-One from within Eclipse, but the Maven build still failed (complained about JustJ version):
> 
> [ERROR] Cannot resolve target definition: [ERROR] Software being installed: org.eclipse.justj.epp.feature.group 21.0.0.v20240807-1553 [ERROR] Missing requirement: org.eclipse.justj.epp 21.0.0.v20240807-1553 requires 'osgi.bundle; org.eclipse.justj.openjdk.hotspot.jre.full [21.0.4.v20240802-1551,22.0.0)' but it could not be found [ERROR] Cannot satisfy dependency: org.eclipse.justj.epp.feature.group 21.0.0.v20240807-1553 depends on: org.eclipse.equinox.p2.iu; org.eclipse.justj.epp [21.0.0.v20240807-1553,21.0.0.v20240807-1553]
> 

I'm not sure which update sites have for that.  These two sites should generally always work:

![image](https://github.com/user-attachments/assets/0ec9215a-7bb8-4d96-b8eb-ab8bdf0bf1dc)

There was just a new release of Java 21 yesterday so that sites have new content today.

> Similar problems occur when one tries to build an older BIRT release from source (at least without a Nexus server).
> 
> My questions are:
> 
>     1. Main question: How can I create a build based on a fixed set of dependencies which can still be built several months later as well?
> 
The final commits at the end of the release cycle strive to do that.

>     2. Is changing the single line in the `*.target` file like I did the right way?
> 
Hopefully yes.  But the target is generated so it's possible (but not very likely) that such a simple change is not sufficient.
> 
> I guess the suffix `-I-builds` means "intermediate builds on the way to the release version", and without the suffix it is the release version.
> 
Yes, that's how the Platform does it.
>     3. What can I do about the JustJ problem in the Maven build?
> 
Use the URLs I showed above.
>     4. Why does Eclipse remove the intermediate repositories at all?
> 
> 
Because all such content is out-dated pre-release content.  Even for BIRT, one nightly build is basically deleted every time there is a new nightly build and all the milestones are deleted when new version is started.  But we do it a little more nicely so that there are stable "latest" repository URLs that will always work.  The Platform does not do it this way.
> If the file system or repo-software that is used internally deduplicates BLOBs, keeping it (instead of deleting it) should be no problem.
> 
All this is just content in a plain old file system served by a plain old http server.
>     5. How do others tackle this?
> 
I've tried to stabilize the target platform at the end of the release so that there is a stable starting/ending point.
>     6. Would it even be possible to have such reliable builds for BIRT releases as well?
> 
I think so, but in the end, that's not my problem because we don't provide maintenance releases.  There's more than enough to do already.
> 
> Note: I'm not even talking about _reproducible_ builds, which would mean that the build artefacts are bit-wise identical everytime they are built.

**Reply by hvbtup** [reply]
Thanks, Ed, for your answers.

> The terms of the EPL require you to make such modifications public.

They are of course: https://gitub.com/triestram-partner/birt

> Note that for the most recent releases I've been stabilizing the target platform fix repository locations

But you overwrote this change with 9d4731cbfe197ed30ffd24d27be5a8b2cf2a8ee0 "Prepare for the 4.18 release":

_The target file contains 4.34-I-builds again!_

> Because I generally expect that no matter when one builds one will want the latest Java release with the latest CVE fixes.

As a default, that expectation is certainly ok.
But sometimes you have to weigh up between the risk of using new softare in general, stability and security fixes.
In the past, we sometimes found incompatible changes in the JRE causing errors in our software which could only be detected at runtime.

> Generally using package imports instead of bundle requirements helps avoid that, ...

Fun fact: Just yesterday, one of our Eclipse RCP experts told me that an external consultant told him exactly this ...

> ... but the code base how it is.

... and I told him that I thought that the BIRT project probably uses best practices. Now I stand corrected. Maybe it was best practice when BIRT was started decades ago.

> ... we don't provide maintenance releases. There's more than enough to do already.

Of course, I know.

Regarding JustJ, I'll try your hint.

**Reply by merks** [reply]
> > Note that for the most recent releases I've been stabilizing the target platform fix repository locations
> 
> But you overwrote this change with [9d4731c](https://github.com/eclipse-birt/birt/commit/9d4731cbfe197ed30ffd24d27be5a8b2cf2a8ee0) "Prepare for the 4.18 release":
> 
> _The target file contains 4.34-I-builds again!_

Yes, of course.  During ongoing release cycle we build against **all the latest dependencies** including the Platform's latest I-builds.  It's only at the end of the cycle that projects produce permanent stable sites.
> 
> > Because I generally expect that no matter when one builds one will want the latest Java release with the latest CVE fixes.
> 
> As a default, that expectation is certainly ok. But sometimes you have to weigh up between the risk of using new softare in general, stability and security fixes. In the past, we sometimes found incompatible changes in the JRE causing errors in our software which could only be detected at runtime.
I don't personally have to weigh so many such risks.  We ship with all the latest and those things should be tested.  

Of course anyone building a fork is able to choose whatever they want to build against. Permanent choices are available here:

- https://download.eclipse.org/justj/epp/
- https://download.eclipse.org/justj/jres/21/updates/

> 
> > Generally using package imports instead of bundle requirements helps avoid that, ...
> 
> Fun fact: Just yesterday, one of our Eclipse RCP experts told me that an external consultant told him exactly this ...
> 
It's considered a best practice.  But then, many projects to version the packages properly or at all, and then it because less useful.  Or there is the fun split-package problem.

> > ... but the code base how it is.
> 
> ... and I told him that I thought that the BIRT project probably uses best practices. Now I stand corrected. Maybe it was best practice when BIRT was started decades ago.
> 
It's moving in the right direction, but there was not so much best practice when I started the releng work.
> > ... we don't provide maintenance releases. There's more than enough to do already.
> 
> Of course, I know.
> 
> Regarding JustJ, I'll try your hint.
Yes, this should be the least of your problems.  Many stable sites to choose from are available and just changing the URLs will do the trick...

**Comment by hvbtup** [comment]
Unfortunately, the Maven build does not work with the recommeded changes for JustJ.

[ERROR] Cannot resolve target definition:
[ERROR]   You requested to install 'org.eclipse.equinox.p2.iu; a.jre.org.eclipse.justj.openjdk.hotspot.jre.full [17.0.0,17.0.0]' but it could not be found

Probably there's more to do, but I don't know what.

**Reply by merks** [reply]
The problem is I don't know what state you are in.  Older Java 17 versions (which is not what the most recent release is using) are here:

https://download.eclipse.org/justj/jres/17/updates/
https://download.eclipse.org/justj/epp/release

**Comment by hvbtup** [comment]
> Yes, of course. During ongoing release cycle we build against all the latest dependencies including the Platform's latest I-builds. It's only at the end of the cycle that projects produce permanent stable sites.

OK.
So if I want to have a stable build, I' should only merge until the latest release tag, and then maybe cherry-pick later commits I think I really need.

My colleague will invest 1-2h to see if we can make the Maven build work. If it succeeds, I'll report here.

**Reply by merks** [reply]
Yes, that's the intent, i.e., that the target platform is based on stable repositories for state of the git repository at the release tag.

If you have problems, I will try to help.

**Comment by rostup** [comment]
Hi. I'm the colleague @hvbtup was talking about. I was experimenting with the target platform today and encountered a strange error:

I had created a new target file in my workspace and wanted to add the necessary repos and dependencies one by one, starting with the JREs. After adding the software site https://download.eclipse.org/justj/jres/17/updates/release/17.0.13, I immediately got an error message in the target editor stating that several dependencies are unsatisfiable. One of them for example is a missing dependency
- **from org.eclipse.justj.openjdk.hotspot.jre.minimal.feature.group 17.0.13.v20241023-1126**
- **to org.eclipse.equinox.p2.iu; org.eclipse.justj.openjdk.hotspot.jre.minimal [17.0.13.v20241023-1126,17.0.13.v20241023-1126]**.

What is wrong here? It sounds like the repo is missing a dependency to itself...

**Reply by merks** [reply]
How might someone else reproduce this problem?  Note that birt is not using jre.minimal so I have not the first clue what exactly you are doing.

**Reply by rostup** [reply]
I created a new blank target definition and added the software site https://download.eclipse.org/justj/jres/17/updates/release/17.0.13 to reproduce the issue. It looks like this:
![image](https://github.com/user-attachments/assets/186aaf0a-44be-46d6-9311-208f934b4135)

**Reply by merks** [reply]
I can't see the definition from this picture.  Moreover, I don't know what you are trying to do nor I why you are trying to do something this way.  I did mention the target is generated so it would seem better that you generate it by changing the setup definition:

![image](https://github.com/user-attachments/assets/7f14d8c2-d5f1-4674-9eca-3f4cf482b0d9)

And using Perform Setup tasks to resolve the targlet definition to regenerate the target platform:

![image](https://github.com/user-attachments/assets/fe96a700-4371-4e22-b121-5aea80251424)

I believe I've also mentioned that for the past two releases the state of the *.target file should be usable to rebuild from the release tag.
